service/remotecfg/remotecfg.go

```diff
@@ -0,0 +1,281 @@
+	}
+}
+
+// SetToDefault implements river.Defaulter.
+func (a *Arguments) SetToDefault() {
+	*a = GetDefaultArguments()
+}
+
+// Validate implements river.Validator.
+func (a *Arguments) Validate() error {
+	// We must explicitly Validate because HTTPClientConfig is squashed and it
+	// won't run otherwise
+	if a.HTTPClientConfig != nil {
+		return a.HTTPClientConfig.Validate()
+	}
+
+	return nil
+}
+
+// Data includes information associated with the remotecfg service.
+type Data struct {
+}
+
+// New returns a new instance of the remotecfg service.
+func New(opts Options) (*Service, error) {
+	basePath := filepath.Join(opts.StoragePath, ServiceName)
+	err := os.MkdirAll(basePath, 0750)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Service{
+		opts:   opts,
+		ticker: time.NewTicker(math.MaxInt64),
+	}, nil
+}
+
+// Data returns an instance of [Data]. Calls to Data are cachable by the
+// caller.
+//
+// Data must only be called after parsing command-line flags.
+func (s *Service) Data() any {
+	return map[string]string{}
+}
+
+// Definition returns the definition of the remotecfg service.
+func (s *Service) Definition() service.Definition {
+	return service.Definition{
+		Name:       ServiceName,
+		ConfigType: Arguments{},
+		DependsOn:  nil, // remotecfg has no dependencies.
+	}
+}
+
+var _ service.Service = (*Service)(nil)
+
+// Run implements [service.Service] and starts the remotecfg service. It will
+// run until the provided context is canceled or there is a fatal error.
+func (s *Service) Run(ctx context.Context, host service.Host) error {
+	s.ctrl = host.NewController(ServiceName)
+
+	// Run the service's own controller.
+	go func() {
+		s.ctrl.Run(ctx)
+	}()
+
+	// We're on the initial start-up of the service.
+	// Let's try to read from the API, parse and load the response contents.
+	// If either the request itself or parsing its contents fails, try to read
+	// from the on-disk cache on a best-effort basis.
+	var (
+		gcr              *connect.Response[agentv1.GetConfigResponse]
+		src              *flow.Source
+		getErr, parseErr error
+	)
+
+	s.mut.RLock()
+	gcr, getErr = s.asClient.GetConfig(ctx, s.getConfigRequest)
+	if getErr == nil {
+		src, parseErr = flow.ParseSource(ServiceName, []byte(gcr.Msg.GetContent()))
+	}
+
+	// Reading from the API succeeded, let's try to load the contents.
+	if getErr == nil && parseErr == nil {
+		err := s.ctrl.LoadSource(src, nil)
+		if err != nil {
+			level.Error(s.opts.Logger).Log("msg", "could not load the API response contents", "err", err)
+		} else {
+			s.currentConfigHash = getHash(gcr.Msg.Content)
+		}
+	} else {
+		// Either the API call or parsing its contents failed, let's try the
+		// on-disk cache.
+		level.Info(s.opts.Logger).Log("msg", "falling back to the on-disk cache")
+		b, err := os.ReadFile(s.dataPath)
+		if err != nil {
+			level.Error(s.opts.Logger).Log("msg", "could not read from the on-disk cache", "err", err)
+		}
+		if len(b) > 0 {
```

Comment:
Should this section still run if there was an error reading the file?
```suggestion
		} else if len(b) > 0 {
```

Reply:
I'm resolving this comment as after some refactoring I believe the logic is more straightforward now.